bin/decode_assoc.R

```diff
@@ -0,0 +1,350 @@
+  }
+
+  num_vars <- 2  # hard-coded for now, since there is --var1 and --var2.
+
+  # Convert strings to integers
+  cohorts <- as.integer(reports$cohort)
+
+  # Hack for Chrome: like AdjustCounts in decode_dist.R.
+  cohorts <- cohorts %% params$m
+
+  # Assume the input has 0-based cohorts, and change to 1-based cohorts.
+  cohorts <- cohorts + 1
+
+  # i.e. create a list of length 2, with identical cohorts.
+  # NOTE: Basic RAPPOR doesn't need cohorts.
+  cohorts_list <- rep(list(cohorts), num_vars)
+
+  # i.e. create a list of length 2, with identical cohorts.
+  string_params <- params
+
+  bool_params <- params
+  # HACK: Make this the boolean.  The Decode() step uses k.  (Note that R makes
+  # a copy here)
+  bool_params$k <- 1
+
+  params_list <- list(bool_params, string_params)
+
+  Log('CreateAssocStringMap')
+  string_map <- CreateAssocStringMap(map, params)
+
+  Log('CreateAssocBoolMap')
+  bool_map <- CreateAssocBoolMap(params)
+
+  map_list <- list(bool_map, string_map)
+
+  string_var <- reports[[opts$var1]]
+  bool_var <- reports[[opts$var2]]
+
+  Log('Preview of string var:')
+  print(head(table(string_var)))
+  cat('\n')
+
+  Log('Preview of bool var:')
+  print(head(table(bool_var)))
+  cat('\n')
+
+  # Split ASCII strings into array of numerics (as required by association.R)
+
+  Log('Splitting string reports (%d cores)', opts$num_cores)
+  string_reports <- mclapply(string_var, function(x) {
+    # function splits strings and converts them to numeric values
+    # rev needed for endianness
+    rev(as.integer(strsplit(x, split = "")[[1]]))
+  }, mc.cores = opts$num_cores)
+
+  Log('Splitting bool reports (%d cores)', opts$num_cores)
+  # Has to be an list of length 1 integer vectors
+  bool_reports <- mclapply(bool_var, function(x) {
+    as.integer(x)
+  }, mc.cores = opts$num_cores)
+
+  reports_list <- list(bool_reports, string_reports)
+
+  Log('Association for %d vars', length(reports_list))
+
+  if (opts$em_executable != "") {
+    Log('Will shell out to %s for native EM implementation', opts$em_executable)
+    em_iter_func <- ConstructFastEM(opts$em_executable, opts$tmp_dir)
+  } else {
+    Log('Will use R implementation of EM (slow)')
+    em_iter_func <- EM
+  }
+
+  em_result <- ComputeDistributionEM(reports_list, cohorts_list, map_list,
+                                     ignore_other = FALSE,
+                                     params_list = params_list,
+                                     marginals = NULL,
+                                     estimate_var = FALSE,
+                                     num_cores = opts$num_cores,
+                                     em_iter_func = em_iter_func,
+                                     max_em_iters = opts$max_em_iters)
+  Log("Association results:")
+  fit <- em_result$fit
+  print(fit)
+
+  pretty <- apply(fit * 100, 1, function(row) sprintf("%.1f%%", row))
+  Log("Percentages:")
+  print(pretty)
+
+  fit_df <- as.data.frame(fit)
+  print(fit_df)
+
```

Comment:
Hey Andy, for now, would you like to use the trick and rename Other to FALSE for boolean so that it is consistent with internal reports?

I tested out the code with the foll. to produce internal reports that are consistent with what I did manually.

rownames(fit_df)[which(rownames(fit_df) == "Other")] = "FALSE"